File: azkar/src/App/CardStack.swift
import AppKit

/// The cards on screen: top-right on the primary display, oldest on top.
@MainActor
final class CardStack {
  private var windows: [CardWindow] = []
  private let margin: CGFloat = 16
  private let gap: CGFloat = 10

  /// The display the cards go on: the primary one (with the menu bar), so they stay put instead
  /// of following keyboard focus between screens the way NSScreen.main does.
  var screen: () -> NSScreen? = { NSScreen.screens.first }

  /// Called whenever a card is removed.
  var onChange: (() -> Void)?

  /// Called when a card closes because its count was reached (not when it's closed with ×).
  var onComplete: (() -> Void)?

  var count: Int { windows.count }

  func push(_ window: CardWindow) {
    window.onClick = { [weak self, weak window] in
      guard let self, let window else { return }
      self.tap(window)
    }
    window.onClose = { [weak self, weak window] in
      guard let self, let window else { return }
      self.remove(window)
    }
    windows.append(window)
    layout(animated: true, fresh: window)
    window.alphaValue = 0
    window.orderFrontRegardless()
    NSAnimationContext.runAnimationGroup { ctx in
      ctx.duration = 0.25
      window.animator().alphaValue = 1
    }
  }

  /// One repetition on `window`; it closes when its count is reached.
  func tap(_ window: CardWindow) {
    guard window.tap() else { return }
    remove(window)
    onComplete?()
  }

  func tapOldest() {
    if let oldest = windows.first { tap(oldest) }
  }

  func dismissOldest() {
    if let oldest = windows.first { remove(oldest) }
  }

  func dismissAll() {
    for window in windows { window.orderOut(nil) }
    windows.removeAll()
    onChange?()
  }

  private func remove(_ window: CardWindow) {
    windows.removeAll { $0 === window }
    window.orderOut(nil)
    layout(animated: true)
    onChange?()
  }

  /// Cards flow down from the top-right corner; if they run out of room they overlap at the bottom.
  func layout(animated: Bool, fresh: CardWindow? = nil) {
    guard let display = screen() else { return }
    let area = display.visibleFrame
    var top = area.maxY - margin
    var frames: [(CardWindow, NSRect)] = []
    for w in windows {
      let size = w.frame.size
      let y = max(top - size.height, area.minY + margin)
      frames.append((w, NSRect(x: area.maxX - margin - size.width, y: y, width: size.width, height: size.height)))
      top = y - gap
    }
    NSAnimationContext.runAnimationGroup { ctx in
      ctx.duration = animated ? 0.2 : 0
      for (w, frame) in frames {
        if animated && w !== fresh {
          w.animator().setFrame(frame, display: true)
        } else {
          w.setFrame(frame, display: true)
        }
      }
    }
  }
}

File: azkar/kotlin/shared/src/commonMain/kotlin/com/wildduck/azkar/app/CardStack.kt
// The cards on screen: what they show, and where they go.
package com.wildduck.azkar.app

import com.wildduck.azkar.core.Card
import com.wildduck.azkar.core.TapCounter
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** One card as it is drawn: its zikr, how far its count has got, and the id its window is tracked by. */
data class CardOnScreen(val id: Long, val card: Card, val badge: String, val fraction: Double)

/** The part of a display cards may use, without the menu bar or taskbar. */
data class CardArea(val left: Int, val top: Int, val right: Int, val bottom: Int)

data class CardSpot(val x: Int, val y: Int)

private const val MARGIN = 16
private const val GAP = 10

/** Cards flow down from the top-right corner; if they run out of room they overlap at the bottom. */
fun cardPositions(area: CardArea, width: Int, heights: List<Int>): List<CardSpot> {
    val x = area.right - MARGIN - width
    var next = area.top + MARGIN
    return heights.map { height ->
        val y = minOf(next, area.bottom - MARGIN - height)
        next = y + height + GAP
        CardSpot(x, y)
    }
}

/** The cards waiting on screen, oldest first. `onComplete` runs when a card's count is reached (not on close). */
class CardStack(private val onComplete: (Card) -> Unit = {}) {
    private class Shown(val id: Long, val card: Card) {
        val counter = TapCounter(card.zikr.count)
    }

    private val shown = mutableListOf<Shown>()
    private var nextId = 1L

    private val _cards = MutableStateFlow<List<CardOnScreen>>(emptyList())
    val cards: StateFlow<List<CardOnScreen>> = _cards.asStateFlow()

    val count: Int get() = shown.size

    /** Adds a card and returns the id to tap or dismiss it by. */
    fun push(card: Card): Long {
        val fresh = Shown(nextId++, card)
        shown += fresh
        publish()
        return fresh.id
    }

    /** One repetition; the card leaves the screen once its count is reached. */
    fun tap(id: Long) {
        val card = shown.firstOrNull { it.id == id } ?: return
        if (!card.counter.tap()) {
            publish()
            return
        }
        shown -= card
        publish()
        onComplete(card.card)
    }

    /** The shortcut counts the card that has been waiting longest. */
    fun tapOldest() {
        shown.firstOrNull()?.let { tap(it.id) }
    }

    /** Closes a card, whatever is left of its count. */
    fun dismiss(id: Long) {
        if (shown.removeAll { it.id == id }) publish()
    }

    fun dismissOldest() {
        shown.firstOrNull()?.let { dismiss(it.id) }
    }

    fun dismissAll() {
        if (shown.isEmpty()) return
        shown.clear()
        publish()
    }

    private fun publish() {
        _cards.value = shown.map { CardOnScreen(it.id, it.card, it.counter.badge, it.counter.fraction) }
    }
}
